truss/contexts/image_builder/cache_warmer.py

```diff
@@ -32,52 +35,105 @@ def _download_from_url_using_b10cp(
     )
 
 
-def split_gs_path(gs_path):
+def split_path(path, prefix="gs://"):
     # Remove the 'gs://' prefix
-    path = gs_path.replace("gs://", "")
+    path = path.replace(prefix, "")
 
     # Split on the first slash
     parts = path.split("/", 1)
 
     bucket_name = parts[0]
-    prefix = parts[1] if len(parts) > 1 else ""
+    path = parts[1] if len(parts) > 1 else ""
 
-    return bucket_name, prefix
+    return bucket_name, path
+
+
+def parse_s3_service_account_file(file_path):
+    # open the json file
+    with open(file_path, "r") as f:
+        data = json.load(f)
+
+    # validate the data
+    if "aws_access_key_id" not in data or "aws_secret_access_key" not in data:
+        raise ValueError("Invalid AWS credentials file")
+
+    # parse the data
+    aws_access_key_id = data["aws_access_key_id"]
+    aws_secret_access_key = data["aws_secret_access_key"]
+    aws_region = data["aws_region"]
+
+    return aws_access_key_id, aws_secret_access_key, aws_region
 
 
 def download_file(
     repo_name, file_name, revision_name=None, key_file="/app/data/service_account.json"
 ):
     # Check if repo_name starts with "gs://"
-    if "gs://" in repo_name:
+    if repo_name.startswith(("gs://", "s3://")):
```

Comment:
[non-blockin] This code is starting to get a little unwieldy. Proposal:

Let's make a `CacheRepository` abstract class, and have subclasses to represent the different kind of repos that you can have here. Something like this:


```python
class RepositoryCache(abc):
    # THis is an abstract class
    def download():
          raise NotImplementedError # use whatever the abc thing is 

    @staticmethod
    def from_repo_id( ) -> "RepositoryCache":
        # Based on repo_id string, return correct subclass
        ...
 class S3Cache(RepositoryCache):

       def download():
            # S3 download logic goes here
            
class GCSCache(RepositoryCache):

class HFCache(RepositoryCahce)
# ... etc

cache = RepositoryCache.from_repo_id(repo_id)
cache.download()
```

Reply:
Yes, I 100% agree. Will implement this.